src/components/order_match/trace.rs

```
@@ -0,0 +1,294 @@
+    core::{
+        backend::{
+            simd::{
+                column::BaseColumn,
+                m31::{PackedBaseField, N_LANES},
+                qm31::PackedSecureField,
+                SimdBackend,
+            },
+            Column,
+        },
+        fields::m31::BaseField,
+        poly::{circle::CircleEvaluation, BitReversedOrder},
+        ColumnVec,
+    },
+};
+use tracing::{debug, span, Level};
+
+use super::{MatchColumn, MatchElements};
+
+/// Preprocessed Trace for Order Matches, each row consisting of a single field element
+/// First row is M31(1), rest are M31(0)
+pub fn preprocessed_trace(
+    log_size: u32,
+) -> ColumnVec<CircleEvaluation<SimdBackend, BaseField, BitReversedOrder>> {
+    vec![IsFirst::new(log_size).gen_column_simd()]
+}
+
+/// Trace for the Order Match Operations, each row consisting of a [BaseField; InstructionColumn::MAIN_COLS]
+/// Generic over the Order Side and Order Match Type
+/// This Function does not check the validity of the instructions
+pub fn trace<S: OrderSide, T: OrderMatchType>(
+    mut matches: Instructions<BaseField>,
+) -> (
+    ColumnVec<CircleEvaluation<SimdBackend, BaseField, BitReversedOrder>>,
+    Claim<MatchColumn<T>>,
+) {
+    let _span = span!(
+        Level::INFO,
+        "Matches: Main Trace",
+        "{} {}",
+        T::NAME,
+        S::NAME,
+    )
+    .entered();
+    // calculate shape of the trace table
+    let log_size = (matches.len() - 1).ilog2() + 1;
+    debug!("Len: {}", matches.len());
+    debug!("Log Size: {}", log_size);
+    // pad matches to a power of 2
+    let mut dummy = matches[0];
+    dummy[InstructionColumn::IS_REAL] = BaseField::zero();
+    for _ in 0..(1 << log_size) - matches.len() {
+        matches.push(dummy);
+    }
+    let mut trace = ComponentTrace::<N_INSTRUCTION_FELTS>::zeroed(log_size);
+    trace
+        .par_iter_mut()
+        .zip(matches.par_chunks_exact(N_LANES))
```

Doesn't `par_chunks_exact(N_LANES)` silently skips rows when `len < 16`
Shouldn't we use `par_chunks` instead of `par_chunks_exact`

---

is the size of matches is 2^n, the minimumm value of m is set to 4, so par_chunks_exact should'nt be a problem